src/lib/zkapp.ts

```diff
@@ -72,28 +70,41 @@ import {
 } from './mina/smart-contract-context.js';
 import { deprecatedToken } from './mina/token/token-methods.js';
 import type { TokenContract } from './mina/token/token-contract.js';
+import { assertPromise } from './util/assert.js';
 
 // external API
 export { SmartContract, method, DeployArgs, declareMethods, Account, Reducer };
 
 const reservedPropNames = new Set(['_methods', '_']);
+type AsyncFunction = (...args: any) => Promise<any>;
 
 /**
- * A decorator to use in a zkApp to mark a method as callable by anyone.
+ * A decorator to use in a zkApp to mark a method as provable.
  * You can use inside your zkApp class as:
  *
  * ```
- * \@method myMethod(someArg: Field) {
+ * \@method async myMethod(someArg: Field) {
+ *   // your code here
+ * }
+ * ```
+ *
+ * To return a value from the method, you have to explicitly declare the return type using the {@link method.returns} decorator:
+ * ```
+ * \@method.returns(Field)
+ * async myMethod(someArg: Field): Promise<Field> {
  *   // your code here
  * }
  * ```
  */
-function method<T extends SmartContract>(
-  target: T & { constructor: any },
-  methodName: keyof T & string,
-  descriptor: PropertyDescriptor
+function method<K extends string, T extends SmartContract>(
+  target: T & {
+    [k in K]: (...args: any) => Promise<void>;
+  },
```

Comment:
the decorator type now enforces a particular function signature of the method that is decorated:
```ts
(...args: any) => Promise<void>
```